File: GAMEOFLIFE/Fox.cpp
#include "Fox.h"
#include "Animal.h"
#include "World.h"

Fox::Fox(int x, int y) : Animal('^', 3, x, y, 3, 7) {
}

void Fox::Action() {

	AddAge();

	std::random_device rd;
    std::mt19937 eng(rd());
    std::uniform_int_distribution<> range(0, 3);
	int move = range(eng);

	pair<int, int> newpos;

	while (move <= 8) {
		newpos = { GetX() + moves[move%4][0], GetY() + moves[move%4][1] };
		if (world->GetOrganism(newpos) == nullptr) {
			if (SetPos(newpos, false))
				break;
		}
		else if(world->GetOrganism(newpos)->GetStrength()<=GetStrength()){
			if (SetPos(newpos, false))
				break;
		}
		else if(world->GetOrganism(newpos)->GetStrength()>GetStrength()){
			world->AddLog(this, "avoided a stronger enemy", 1);
		}
		move++;
	}
}

Fox* Fox::Child() {
	return new Fox(*this);
}

File: GAMEOFLIFEJAVA/gameoflife/organisms/animals/Fox.java
package gameoflife.organisms.animals;

import gameoflife.structs.Vector;
import gameoflife.structs.RandomGenerator;

public class Fox extends Animal {
    
    public Fox(Vector position) {
        super("^", 3, 7, position);
    }

    @Override
    public void Action() {

        RandomGenerator r = new RandomGenerator();

        AddAge();

        Vector[] allowedMoves = world.GetPossibleMoves();
        for(int move = r.random(0,allowedMoves.length-1); move<=world.GetPossibleMoves().length*2; move++){
            if(!world.OnBoard(GetPos().Add(allowedMoves[move%world.GetPossibleMoves().length])))
                continue;
            if(world.GetOrganism(GetPos().Add(allowedMoves[move%world.GetPossibleMoves().length]))==null){
                SetPos(GetPos().Add(allowedMoves[move%world.GetPossibleMoves().length]));
                return;
            }
            if(world.GetOrganism(GetPos().Add(allowedMoves[move%world.GetPossibleMoves().length])).GetStrength()<=GetStrength()){
                SetPos(GetPos().Add(allowedMoves[move%world.GetPossibleMoves().length]));
                return;
            }
            else{
                world.Log("Fox avoided a stronger enemy");
            }
        }
    }
}
